### PR TITLE
[WIP] update: only display updated installed formulae.

### PR DESCRIPTION
**Pull request by mikemcquaid**
Chances are people aren't interested in modifications to formulae that aren't installed but we've modified in the same way they might be with new or deleted formulae.

Also, print the versions of these formulae too.

CC @Homebrew/owners for review. Not done yet as I wrote it on a :airplane: so not tested.

### CODE REVIEW

**Comment by chdiza** [comment]
:-1: I always want a list of everything that's changed.  For instance...

Sometimes I'm waiting on some uninstalled formula `foo` to get a version bump (or some other fix) before I install it.  The presence of `foo` in the output of `update` tells me that change probably just landed, and doesn't require me to do anything beyond what I would normally do anyway, namely `update`.

Sometimes I catch an outdated formula `bar` and start working on a PR.  The first thing I do is to `update` on master to make sure the PR won't need rebasing.  Sometimes the output of `update` lists `bar`, which prevents me from wasting my time and making a duplicate PR.

Finally, the output of `update` is very often how I discover projects.  I see the name in the output and think "I wonder what *that* is", and then I `brew home` it.  Sometimes I like what I see.  Seven of my brewed formulas were brought to my attention this way.  The "free advertising" role of `update` seems important to me.

Has anyone actually requested the functionality in the PR?  If not, I don't see a reason to merge it.  Or, you could tuck it behind a flag, e.g., `brew update --only-installed`.

**Comment by chdiza** [comment]
I thought of another reason not to merge this PR (at least not without hiding it behind a flag).

I have custom formulas that sometimes have a different name than their official counterparts.  I know other people do this in taps sometimes to avoid problems (and this is even recommended sometimes by maintainers).

Suppose, e.g., that I have a custom `chdiza-gnupg2.rb` formula.  Then AFAICT from the code in the PR, `brew update` will NOT notify my if the official `gnupg2` gets modified, because `Formula.factory()` will find no such thing as `chdiza-gnupg2`.  This means I won't notice that `gnupg2` has received some change, perhaps a version bump or other fix.

**Comment by mikemcquaid** [comment]
> I always want a list of everything that's changed. For instance...

Sure, I just don't think your case is the usual one for Homebrew. 

> Has anyone actually requested the functionality in the PR? If not, I don't see a reason to merge it.

https://github.com/Homebrew/homebrew/pull/25951, https://github.com/Homebrew/homebrew/issues/27217, https://github.com/Homebrew/homebrew/issues/35468 all seem to touch on a similar problem.

> Or, you could tuck it behind a flag, e.g., brew update --only-installed.

More options suck for usability (particularly when you're trying to make a change to improve usability).

**Comment by chdiza** [comment]
None of those PRs ask that uninstalled formulas have their modification status go totally unreported.  

Only #35468 is even relevant; the others are asking for *additional* output, not that currently-produced information be suppressed.  And even in the case of #35468, uninstalled formulas would *still* get reported as modified as long as their changes were "substantial"; it would treat uninstalled formulas just like it treats installed ones.

In short, I still can't see that anyone is asking for this.

But if we're going to fix what 'aint broke, I would be satisfied if the *current* behavior were tucked behind a flag.  If you're right that this change would be popular and I'm the weird case, then most people wouldn't have to worry about a flag.

File: Library/Homebrew/cmd/update.rb [hunk 1]
```diff
@@ -275,12 +276,15 @@ def select_formula key
       else
         path.basename(".rb").to_s
       end
-    end.sort
+    end.sort.uniq
```
Comment (jacknagel) [inline Library/Homebrew/cmd/update.rb]: Where are duplicates being introduced?

Comment (mikemcquaid) [inline Library/Homebrew/cmd/update.rb]: Whoops, leftover from testing.

File: Library/Homebrew/cmd/update.rb [hunk 2]
```diff
@@ -251,7 +252,7 @@ def dump
     # Key Legend: Added (A), Copied (C), Deleted (D), Modified (M), Renamed (R)
 
     dump_formula_report :A, "New Formulae"
-    dump_formula_report :M, "Updated Formulae"
+    dump_formula_report :M, "Upgraded Formulae"
```
Comment (tdsmith) [inline Library/Homebrew/cmd/update.rb]: Sorry to bikeshed but I like `Updated` here to avoid implying that `update` has any functional overlap with `upgrade`.

File: Library/Homebrew/cmd/update.rb [hunk 3]
```diff
@@ -275,12 +276,15 @@ def select_formula key
...
         path.basename(".rb").to_s
       end
-    end.sort
+    end.sort.uniq
   end
 
   def dump_formula_report key, title
     formula = select_formula(key)
+    formula.map! {|f| Formula.factory(f) rescue nil }
+    formula.select! {|f| f && f.installed? }
```
Comment (tdsmith) [inline Library/Homebrew/cmd/update.rb]: Should this be conditional on `key == :M` (or do you want to stop showing new packages?)

Comment (mikemcquaid) [inline Library/Homebrew/cmd/update.rb]: Yes, good point.

File: Library/Homebrew/cmd/update.rb (hunk 2 above)
Comment (mikemcquaid) [inline Library/Homebrew/cmd/update.rb]: :+1:

**Comment by mikemcquaid** [comment]
Do any @Homebrew/owners have objections to this approach? I'm thinking it may be a good solution to being able to merge the ~3000 formulae descriptions without destroying `brew update` readability (and handling similar cases in future).

Another approach if it's preferred would be to only display formulae that have been upgraded e.g. the formula (or bottle) revision has changed since before the `brew update`. Alternatively we could combine the two and show only installed formulae that have been upgraded. Thoughts?

**Comment by DomT4** [comment]
> Another approach if it's preferred would be to only display formulae that have been upgraded e.g. the formula (or bottle) revision has changed since before the brew update.

:heart_eyes_cat: :heart_eyes_cat: :heart_eyes_cat:. Would make fixing things like plaintext urls/dead links/trivial fixes/etc a lot easier without users seeing the formula has been updated and wondering why they haven't received said update, etc. Would be pretty great.

**Comment by xu-cheng** [comment]
Actually I'm think another approach that it's to only display formulae with version changed.

**Comment by mikemcquaid** [comment]
Closing in favour of https://github.com/Homebrew/homebrew/pull/38817.
